### ISSUE TITLE
Bug for calling end() of SessionRunHook

### ISSUE THREAD

**Issue by byshiue** [tensorflow/tensorflow#29772]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- Linux Ubuntu 18.04
- TensorFlow installed from: pip
- TensorFlow version (use command below): 1.13.1
- Python version: 3.6.7
- CUDA/cuDNN version: CUDA 10.1, cuDNN 7.6
- GPU model and memory: 2080Ti

**Describe the current behavior**
I define a hook class which is inherited from SessionRunHook, and add an assign operation in the begin() function. Then, I run the assign operation in the end() function. Although the value is changed, the results do not store into the checkpoint. So, If I restore the model, I get the model which is not changed. 

**Describe the expected behavior**
Since we only know the end() is called before closing the session. I think the changed result should be store in the chekpoint. Otherwise, the document should provide the ordering of executing all hooks. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
class NetworkPruningHook(tf.train.SessionRunHook):
    def __init__(self):

    def begin(self):
      self.mask_var = []
      ### collect the masked variables
      for v in tf.global_variables():
        if v.name.find("mask:0") != -1:
          self.mask_var.append(v)

      self.check_mask = tf.assign(self.mask_var[0], np.zeros(shape=self.mask_var[0].shape))

    def end(self, session):
      session.run(self.check_mask) # assign 0 to the variable 
      print(session.run(self.mask_var[0])) # get [0, ..., 0]
      
      ### However, if restoring the checkpoint, the self.mask_var[0] is not [0, ..., 0]

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by ravikyram** [comment]
In order to expedite the trouble-shooting process, please provide a full code snippet to reproduce the issue reported here. Thanks!

**Comment by byshiue** [comment]
class NetworkPruningHook(tf.train.SessionRunHook):
     def __init__(self):
        super(NetworkPruningHook, self).__init__()
        self.sparsity = 0
        self.total_step = 0
        self.step = 0
        self.hasInitialize = False

    def initialize(self, sparsity, total_step):
      self.step = 0
      self.total_step = (int)(total_step)
      self.sparsity = sparsity
      self.hasInitialize = True

    def begin(self):
      tf.logging.info("[INFO] sparsity: %d", self.sparsity)
      self.mask_var = []
      self.masked_output = []
      self.assign_ops = []
      self.placeholders = []
      self.scores = []
      self.loss = None
      assert(self.hasInitialize == True)

      ### collect the masked variables
      for v in tf.global_variables():
        if v.name.find("mask:0") != -1:
          self.mask_var.append(v)
          p = tf.placeholder(tf.float32, shape=v.shape)
          self.placeholders.append(p)
          self.assign_ops.append(tf.assign(v, p))
          self.scores.append(np.zeros(v.shape))

      ### collect the outputs of masked layers and loss layer
      for op in tf.get_default_graph().get_operations():
        for t in op.values():
          if t.name.find("mask-output") != -1:
            self.masked_output.append(t)
          elif t.name.find("loss/Mean:0") != -1:
          	self.loss = t

      self.var_grad = tf.gradients(self.loss, self.mask_var)

    def before_run(self, run_context):
      return tf.train.SessionRunArgs([self.var_grad, self.masked_output])

    def after_run(self, run_context, run_values):
      for i in range(len(self.mask_var)):
        score = abs(np.sum(run_values.results[0][i] * run_values.results[1][i], axis=0))
        score_sum = np.sum(score)
        if(score_sum) != 0:
          score = score / score_sum
        self.scores[i] = self.scores[i] + score

      self.step += 1
      if self.step >= self.total_step:
        ### check the zero_density
        zero_density = [np.sum(x)*1. / np.shape(x)[0] for x in run_context.session.run(self.mask_var)]
        average = np.average(zero_density)
        tf.logging.info("zero_density of all layers before pruning: ")
        # for i in range(len(zero_density)):
        #   tf.logging.info("  zero_density of layer %d: %f", i, zero_density[i])
        tf.logging.info("The average zero_density of network: %f", average)

        ### pruning
        threshold = np.percentile([y for x in self.scores for y in x], self.sparsity)
        for i in range(len(self.mask_var)):
          mask = (self.scores[i] > threshold)
          run_context.session.run(self.assign_ops[i], feed_dict={self.placeholders[i]: mask})

        ### check the zero_density
        zero_density = [np.sum(x)*1. / np.shape(x)[0] for x in run_context.session.run(self.mask_var)]
        average = np.average(zero_density)
        tf.logging.info("zero_density of all layers: ")
        # for i in range(len(zero_density)):
        #   tf.logging.info("  zero_density of layer %d: %f", i, zero_density[i])
        tf.logging.info("The average zero_density of network: %f", average)

    def end(self, session):
    	self.hasInitialize = False


Above is my hook, and I add a masked layer by tf.Variable with name "xxx-mask" on the fully-connected layer. If I move the assign operation from begin() to end(), then the assign results will not be written in checkpoint file after training.

**Comment by ravikyram** [comment]
I tried reproducing the issue. But i am not able to get saved checkpoints.Please,help us to give  more information to reproduce the issue.Thanks!

**Comment by byshiue** [comment]
I use the code of https://github.com/google-research/bert directly, I think the estimator will save checkpoints every some steps and when the training is end

**Comment by ravikyram** [comment]
Will it be possible to send me the full minimal code snippet to reproduce the issue for faster resolution.Thanks!

**Comment by ravikyram** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
